# Werkcollege-opdrachten Week 1.3

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige werkcollege de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [2]:
import pandas as pd
import sqlite3
import warnings
warnings.simplefilter('ignore')

Zet de volgende bestanden in een makkelijk te vinden map, als je dat nog niet gedaan hebt:
- go_sales.sqlite
- go_crm.sqlite
- go_staff.sqlite

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige werkcollege de code om een connectie met het bestand go_sales.sqlite te maken.

In [8]:
conn = sqlite3.connect("./go_sales.sqlite")

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- returned_item
- return_reason

In [ ]:
select_all_product = "SELECT * FROM product"
select_all_product_type = "SELECT * FROM product_type"
select_all_product_line = "SELECT * FROM product_line"
select_all_sales_staff = "SELECT * FROM sales_staff"
select_all_sales_branch = "SELECT * FROM sales_branch"
select_all_retailer_site = "SELECT * FROM retailer_site"
select_all_country = "SELECT * FROM country"
select_all_order_header = "SELECT * FROM order_header"
select_all_order_details = "SELECT * FROM order_details"
select_all_returned_item = "SELECT * FROM returned_item"
select_all_return_reason = "SELECT * FROM return_reason"

#tabel: product
#product = pd.read_sql(select_all_product, conn)
#product

#tabel: product_type.
#product_type = pd.read_sql(select_all_product_type, conn)
#product_type

#tabel: product_line
#product_line = pd.read_sql(select_all_product_line, conn)
#product_line

#tabel: sales_staff
sales_staff = pd.read_sql(select_all_sales_staff, conn)
sales_staff

#tabel: sales_branch
#sales_branch = pd.read_sql(select_all_sales_branch, conn)
#sales_branch

#tabel: retailer_site
#retailer_site = pd.read_sql(select_all_retailer_site, conn)
#retailer_site

#tabel: country
#country = pd.read_sql(select_all_country, conn)
#country

#tabel: order_header
#order_header = pd.read_sql(select_all_order_header, conn)
#order_header

#tabel: order_details
#order_details = pd.read_sql(select_all_order_details, conn)
#order_details

#tabel: returned_item
#returned_item = pd.read_sql(select_all_returned_item, conn)
#returned_item

#tabel: return_reason
#return_reason = pd.read_sql(select_all_return_reason, conn)
#return_reason


SALES_STAFF_CODE FIRST_NAME  LAST_NAME                   POSITION_EN  \
0                   4      Denis       Pagé                Branch Manager   
1                   5  Élizabeth     Michel  Level 3 Sales Representative   
2                   6      Émile   Clermont  Level 1 Sales Representative   
3                   7    Étienne     Jauvin  Level 2 Sales Representative   
4                  12    Elsbeth  Wiesinger  Level 2 Sales Representative   
..                ...        ...        ...                           ...   
97                120      Giele   Laermans  Level 1 Sales Representative   
98                121   François    De Crée  Level 1 Sales Representative   
99                122     Yvette    Lattrez  Level 3 Sales Representative   
100               123      Willi  Seefelder  Level 2 Sales Representative   
101               124     Sabine     Grüner  Level 3 Sales Representative   

            WORK_PHONE  EXTENSION                FAX                   EMAIL  \
0    +33 1 68 94 52 20      325.0  +33 1 68 94 56 60       DPage@grtd123.com   
1    +33 1 68 94 52 20      336.0  +33 1 68 94 56 60     EMichel@grtd123.com   
2    +33 1 68 94 52 20      378.0  +33 1 68 94 56 60   EClermont@grtd123.com   
3    +33 1 68 94 52 20      398.0  +33 1 68 94 56 60     EJauvin@grtd123.com   
4    +(49) 40 663 1990     1818.0  +(49) 40 663 4571  EWiesinger@grtd123.com   
..                 ...        ...                ...                     ...   
97     +32 16 20.73.21     1340.0    +32 16 20.73.32   GLaermans@grtd123.com   
98     +32 16 20.73.21     1642.0    +32 16 20.73.32     FDecree@grtd123.com   
99     +32 16 20.73.21     1633.0    +32 16 20.73.32    YLattrez@grtd123.com   
100  +(43) 13 79 56 32      325.0  +(43) 13 79 56 33  WSeefelder@grtd123.com   
101  +(43) 13 79 56 32      348.0  +(43) 13 79 56 33     SGruner@grtd123.com   

     DATE_HIRED  SALES_BRANCH_CODE  
0    1996-11-03                  6  
1    1995-06-08                  6  
2    1998-04-07                  6  
3    1997-08-16                  6  
4    1997-02-27                 13  
..          ...                ...  
97   1999-11-15                 38  
98   1999-12-01                 38  
99   1999-12-09                 38  
100  1998-10-28                 39  
101  1998-02-18                 39  

[102 rows x 10 columns]

Krijg je een "no such table" error? Dan heb je misschien met .connect() per ongeluk een leeg  databasebestand (.sqlite) aangemaakt.<br> <u>Let op:</u> lees eventueel de informatie uit het vorige werkcollege nog eens goed door.

Als je tijdens onderstaande opdrachten uit het oog verliest welke tabellen er allemaal zijn, kan je deze Pythoncode uitvoeren:

In [ ]:
sql_query = "SELECT * FROM sqlite_master WHERE type='table';"
#Vul dit codeblok verder in
pd.read_sql(sql_query, conn)
#Op de puntjes hoort de connectie naar go_sales óf go_staff óf go_crm te staan.

type           name       tbl_name  rootpage  \
0   table        country        country       480   
1   table  order_details  order_details        11   
2   table   order_header   order_header        24   
3   table   order_method   order_method        27   
4   table        product        product        28   
5   table   product_line   product_line        17   
6   table   product_type   product_type        10   
7   table  retailer_site  retailer_site        18   
8   table  return_reason  return_reason        30   
9   table  returned_item  returned_item        31   
10  table   sales_branch   sales_branch         2   
11  table    sales_staff    sales_staff         3   

                                                  sql  
0   CREATE TABLE "country" (\n\t"COUNTRY_CODE"\tIN...  
1   CREATE TABLE "order_details" (\n\t"ORDER_DETAI...  
2   CREATE TABLE "order_header" (\n\t"ORDER_NUMBER...  
3   CREATE TABLE "order_method" (\n\t"ORDER_METHOD...  
4   CREATE TABLE "product" (\n\t"PRODUCT_NUMBER"\t...  
5   CREATE TABLE "product_line" (\n\t"PRODUCT_LINE...  
6   CREATE TABLE "product_type" (\n\t"PRODUCT_TYPE...  
7   CREATE TABLE "retailer_site" (\n\t"RETAILER_SI...  
8   CREATE TABLE "return_reason" (\n\t"RETURN_REAS...  
9   CREATE TABLE "returned_item" (\n\t"RETURN_CODE...  
10  CREATE TABLE "sales_branch" (\n\t"SALES_BRANCH...  
11  CREATE TABLE "sales_staff" (\n\t"SALES_STAFF_C...

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [33]:
#1. Lees het dataFrame in
#2. We willen alles selecteren van deze tabel.
#3. We maken een boolean aan en kijken in de kolom 'PRODUCT_COST' 
#4. Waarin we productiekosten < 100 & > 50 willen projecteren op het scherm

product = pd.read_sql_query(select_all_product, conn)

product_kosten_hoger_lager = ((product['PRODUCTION_COST'] > 50) & (product['PRODUCTION_COST'] < 100))

product.loc[(product_kosten_hoger_lager), ['PRODUCT_NAME', 'PRODUCTION_COST']]






PRODUCT_NAME  PRODUCTION_COST
5          TrailChef Deluxe Cook Set            85.11
7             TrailChef Double Flame            75.00
16                   Hibernator Lite            60.00
17                        Hibernator            86.00
19   Hibernator Self - Inflating Mat            78.55
22               Hibernator Camp Cot            65.33
23      Canyon Mule Climber Backpack            62.50
45           Granite Climbing Helmet            52.86
47             Husky Harness Extreme            55.00
57                    Granite Hammer            57.13
68                         Polar Ice            73.33
69                      Polar Sports            80.00
70                        Polar Wave            60.00
71                     Polar Extreme            72.50
74                      Edge Extreme            80.00
77                         Seeker 35            79.19
78                         Seeker 50            92.58
79                    Seeker Extreme            94.12
82                    Glacier Deluxe            56.25
83                       Glacier GPS            78.55
109                Blue Steel Putter            60.00
110            Blue Steel Max Putter            81.80
113              Course Pro Golf Bag            80.00

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [40]:
#1. Lees de query in om een variabele te maken
#2. Maak een boolean waar marge < 20% of hoger dan 60%
#3. gebruik .loc[(boolean), ['kolommen']]

product = pd.read_sql(select_all_product, conn)

marge_hoger_lager = ((product['MARGIN'] < 0.2)) | ((product['MARGIN'] > 0.6))

product.loc[marge_hoger_lager, ['PRODUCT_NAME', 'MARGIN']]

PRODUCT_NAME  MARGIN
20                Hibernator Pad    0.17
23  Canyon Mule Climber Backpack    0.17
85             BugShield Natural    0.67
86               BugShield Spray    0.67
87         BugShield Lotion Lite    0.70
88              BugShield Lotion    0.63
89             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [ ]:
#1. Lees de query om een variabele te maken
#2. Ik neem aan de we het tabel land gaan inlezen
#3. Maak een boolean met == 'francs'
#4. variabele.loq[bool, [kolom]]

land = pd.read_sql(select_all_country, conn)

land_betaald_met_francs = ((land['CURRENCY_NAME']) == 'francs')

land.loc[land_betaald_met_francs, ['COUNTRY']]



COUNTRY
0        France
7   Switzerland
15      Belgium

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [ ]:
# Lees de query om een variabele aan te maken
# Maak een boolean waarin je marge met meer dan 50% selecteert.
# Kolommen 'INTRODUCTION_DATE' en 'MARGIN'
# Sinds we verschillende waardes willen moeten we drop_duplicates gebruiken

product = pd.read_sql(select_all_product, conn)

marge_meer_dan_vijftig = ((product['MARGIN']) > 0.5)

product.loc[marge_meer_dan_vijftig, ['INTRODUCTION_DATE']].drop_duplicates('INTRODUCTION_DATE')

INTRODUCTION_DATE
66         2000-10-26
85         1995-02-15
101        2003-12-10
103        2003-12-18
104        2003-12-27
106        2004-01-13
110        2003-12-15

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 7 rijen)

In [ ]:
# Lees de query om een variabele aan te maken
# Welk tabel? sales_branch
# Welk kolommen heb je nodig?  ADRESS1, ADRESS2, REGION, CITY
# Maak een boolean waarin ADRESS2 != NaN en REGION NaN
# Dit doen we met notna()

salesbranch = pd.read_sql(select_all_sales_branch, conn)

alle_adressen_bekend = ((salesbranch['ADDRESS2'].notna()) & (salesbranch['REGION'].notna()))

salesbranch.loc[alle_adressen_bekend, ['ADDRESS1', 'CITY']]



ADDRESS1              CITY
2                Singelgravenplein 4         Amsterdam
13  Prol. Paseo de la Reforma No. 51  Distrito Federal
14               202-2-3 Hyakunincho             Tokyo
15                     543-225 Asahi        Osaka City
16                  2315 Queen's Ave         Melbourne
18             Avenida Paulista, 333         São Paulo
24                    3 Albany Court        Birmingham

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [ ]:
#Lees de query om een variabele aan te maken
# Welk tabel hebben we nodig? #country
# Welke kolommen hebben we nodig? #CURRENCY_NAME #COUNTRY
# Maak een boolean waar de landen met dollars of new dollar betalen.
# Sorteer de uitkomst op land. .sort_values

country = pd.read_sql(select_all_country, conn)

country_dollar_new_dollar = ((country['CURRENCY_NAME'] == 'dollars') | (country['CURRENCY_NAME'] == 'new dollar'))

country.loc[country_dollar_new_dollar, ['COUNTRY']].sort_values('COUNTRY')


COUNTRY
14      Australia
3          Canada
11         Taiwan
2   United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [16]:
# Lees de query.
# Welk tabel? Retailer_Site? 
# Welke kolommen? ADDRESS1, ADDRESS2, CITY, POSTAL_ZONE
# Maak een boolean postcode begint met 'd' filter op vestiging die twee adressen hebben.
# Uitkomst, beide adressen en de stad.

vestiging = pd.read_sql(select_all_retailer_site, conn)

postcode_begint_met_d = ( (vestiging['ADDRESS1'].notna()) & (vestiging['ADDRESS2'].notna()) & (vestiging['POSTAL_ZONE'].str[0] == 'D') )

vestiging.loc[postcode_begint_met_d, ['ADDRESS1', 'ADDRESS2', 'CITY']]

ADDRESS1       ADDRESS2       CITY
31  Röntgenstraße 90  3. Tür rechts  Frankfurt
59   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [28]:
# Lees de query
# Welk tabel? returned_item
# COUNT het totaal aantal producten dat is teruggebracht.
# Hoe ? count(), hoe krijg ik één waarde terug? denk door één kolom uittelezen
# Uitkomst totaal aantal producten

producten = pd.read_sql(select_all_returned_item, conn)

totaal_aantal_producten = ( (producten['RETURN_CODE'].count()) )
totaal_aantal_producten




np.int64(619)

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [10]:
# Lees de query in 
# Welk tabel? retailer_site
# Welke kolommen? REGION
# Maak een boolean waarin staat dat een regio notna is.
# var = bool['...'].count
# Uitkomst in 1 waarde

retailer_sites = pd.read_sql(select_all_retailer_site, conn)

not_null_regios = ( (retailer_sites['REGION'].nunique()) )

not_null_regios

110

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [ ]:
# Lees de query in
# import numpy as np?
# Welke tabel? product
# Welke kolommen? MARGIN
# Maak een boolean met de laagste
# Maak een boolean met de hoogste
# Maak een boolean met het gemiddelde.
# Zet die om in een data frame.

product = pd.read_sql_query(select_all_product, conn)

laagste_margin = ((product['MARGIN'].min()))

hoogste_margin = ((product['MARGIN'].max()))

gemiddelde_margin = round(product['MARGIN'].mean(), 2)

resultaat = pd.DataFrame({'Laagste_marge' : [laagste_margin], 'Hoogste_marge' : [hoogste_margin], 'Gemiddelde_marge' : [gemiddelde_margin]})

resultaat



Laagste_marge  Hoogste_marge  Gemiddelde_marge
0           0.17            0.7               0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [22]:
# Lees de query in
# Welk tabel? sales_branch
# Welk kolommen? ADRESS2
# Maak een boolean waar het 2e adres niet bekend is
# Tel deze op, en print het uit.

adres2_niet_bekend = pd.read_sql_query(select_all_sales_branch, conn)

aantal_adressen_niet_bekend =  adres2_niet_bekend['ADDRESS2'].isna().sum()

int_aantal_adressen_niet_bekend = int(aantal_adressen_niet_bekend)

int_aantal_adressen_niet_bekend




19

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [29]:
# Lees de query in
# Welke tabel ? order_details
# Welke kolom ? 'UNIT_COST, UNIT_PRICE, UNIT_SALE_PRICE
# Maak een boolean met avg kostprijs van verkochte producten
# Waarop korting (unit_sale_price < unit_price) is verleend ?
# Print alleen de avg uit.

verkochte_producten = pd.read_sql_query(select_all_order_details, conn)

korting = verkochte_producten['UNIT_SALE_PRICE'] < verkochte_producten['UNIT_PRICE']

# neem alleen de rijen met die korting.
kosten_met_korting = verkochte_producten[korting]['UNIT_COST']

#neem dan het gemiddelde van die rijen.
gemiddelde_kostprijs = round(kosten_met_korting.mean(), 2)

float_gemiddelde_kostprijs = float (gemiddelde_kostprijs)

float_gemiddelde_kostprijs

127.34

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [31]:
# Lees de query in
# Welk tabel? sales_staff
# Welke kolommen? POSITION_EN
# Aantal medewerkers per medewerkersfunctie, denk met group by ?
# Tel het op per group.
# Print het uit.

medewerker = pd.read_sql_query(select_all_sales_staff, conn)

overzicht = medewerker.groupby('POSITION_EN').size().reset_index(name='AANTAL_MEDEWERKERS')

overzicht


POSITION_EN  AANTAL_MEDEWERKERS
0                Branch Manager                  19
1              District Manager                   4
2               General Manager                   1
3  Level 1 Sales Representative                  13
4  Level 2 Sales Representative                  39
5  Level 3 Sales Representative                  24
6              Regional Manager                   2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [34]:
# Lees de query in
# Per telefoon nummer, aantal medewerkers.
# Toon allen de telefoonnummers waarom meer dan 4 medewerkers bereikbaar zijn.
# Welk tabel? sales_staff
# Welke kolommen? FAX

medewerker = pd.read_sql_query(select_all_sales_staff, conn)

overzicht = (medewerker.groupby('FAX').size().reset_index(name= 'AANTAL_MEDEWERKERS'))

overzicht_meer_dan_4_medewerkers = overzicht[overzicht['AANTAL_MEDEWERKERS'] > 4]

overzicht_meer_dan_4_medewerkers

FAX  AANTAL_MEDEWERKERS
4     +(44) 121 3583843                   5
10  +(61) 03 2982  4243                   5
11  +31 (0)20 692 93 06                   5
13    +33 1 68 94 56 60                   5
18     1 (206) 292-3312                   5
19     1 (305) 557-7650                   6
20     1 (310) 281-9087                   5
21     1 (403) 232-5831                   5
24     1 (617) 268-3464                   6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [51]:
# Lees de query in
# Welke tabellen? COUNTRY: RETAILER_SITE
# Welke kolommen? 
# 'PK: COUNTRY_CODE, COUNTRY' 
# ADDRESS1, CITY FK: COUNTRY_CODE
# Overzicht : EERSTE ADRES, STAD
# 'where' land = NETHERLANDS.

country = pd.read_sql_query(select_all_country, conn)

retailer_site = pd.read_sql_query(select_all_retailer_site, conn)

overzicht = pd.merge(country, retailer_site, left_on = 'COUNTRY_CODE', how = 'inner', right_on = 'COUNTRY_CODE')

land_is_netherlands = overzicht['COUNTRY'] == 'Netherlands'

resultaat = overzicht.loc[land_is_netherlands, ['ADDRESS1', 'CITY']]

resultaat





ADDRESS1          CITY
215                   Beets 36         Beets
216               Perestraat 1      Deventer
217  Laan van Meerdervoort 966      Den Haag
218             Kalfjeslaan 47    Amstelveen
219               Startbaan 11    Amstelveen
220      Noorderspoorsingel 28     Groningen
221              Hogehilweg 13     Amsterdam
222             Spoorstraat 43    Varsseveld
223             Westhavenweg 8     Amsterdam
224              Westzeedijk 4        Muiden
225            Gooiergracht 91         Laren
226            Hoofdstraat 150   Pieterburen
227                Olde Diek 2         Laren
228     Amsterdamsestraatweg 8      Maarssen
229        Krayenhoffstraat 11     Amsterdam
230         Overschiestraat 19     Amsterdam
231              Havenbuurt 12        Marken
232               Duivendaal 1    Wageningen
233              Voorstraat 37  Nieuweschans
234         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [57]:
# Tabellen? PRODUCT, PRODUCT_TYPE
# Kolommen? 
# PRODUCT : PRODUCT_NAME FK: PRODUCT_TYPE_CODE
# PRODUCT_TYPE: PRODUCT_TYPE_EN PK: PRODUCT_TYPE_CODE
# Lees de query's 
# Join de twee tabellen.
# Maak een boolean waar ProductType == 'Eyewear'
# print tabel uit.

product = pd.read_sql_query(select_all_product, conn)

product_type = pd.read_sql_query(select_all_product_type, conn)

overzicht = pd.merge(product, product_type, left_on = 'PRODUCT_TYPE_CODE', how = 'inner', right_on = 'PRODUCT_TYPE_CODE')

product_type_eyewear = overzicht['PRODUCT_TYPE_EN'] == 'Eyewear'

resultaat = overzicht.loc[product_type_eyewear, ['PRODUCT_NAME']]

resultaat



PRODUCT_NAME
67      Polar Sun
68      Polar Ice
69   Polar Sports
70     Polar Wave
71  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [ ]:
# Welke tabellen? ORDER_HEADER, SALES_STAFF, RETAILER_SITE
# Welke kolommen? 
# ORDER_HEADER = PK: ORDER_NUMBER, FK: RETAILER_SITE_CODE, FK : SALES_STAFF_CODE
# SALES_STAFF = PK: SALES_STAFF_CODE, FIRST_NAME, LAST_NAME, POSITION_EN
# RETAILER_SITE = PK: RETAILER_SITE_CODE, ADDRESS1
# SALES_STAFF -> ORDER_HEADER -> RETAILER_SITE
# Lees de query's in
# Maak een boolean met de naam is 'Branch Manager'
# Merge de tabellen
# Tabel, heeft ADDRES1 van klantvestiging en voornaam, achternaam verkoper.

order_header = pd.read_sql_query(select_all_order_header, conn)
sales_staff = pd.read_sql_query(select_all_sales_staff, conn)
retailer_site = pd.read_sql_query(select_all_retailer_site, conn)

verkoper_is_branch_manager = sales_staff[sales_staff['POSITION_EN'] == 'Branch Manager']

overzicht1 = verkoper_is_branch_manager.merge(order_header, left_on = 'SALES_STAFF_CODE', how='inner', right_on = 'SALES_STAFF_CODE')

overzicht2 = overzicht1.merge(retailer_site, left_on = 'RETAILER_SITE_CODE', how = 'inner', right_on = 'RETAILER_SITE_CODE')

resultaat = overzicht2[['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates()

resultaat


ADDRESS1 FIRST_NAME LAST_NAME
0  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 7 rijen) 


In [84]:
# Welke tabellen?  SALES_STAFF, ORDER_HEADER
# Welke kolommen? 
# SALES_STAFF: PK: SALES_STAFF_CODE, POSITION_EN like 'Manager'.
# ORDER_HEADER : PK: ORDER_NUMBER, ORDER_DATE?? FK: SALES_STAFF_CODE
# Hoe joinen we deze?
# PK -> FK : SALES_STAFF_CODE
# Lees de query's in
# Maak een boolean van de verkopers met het woord 'Manager' in hun positie naam (...)
# Merge de tabellen één voor één.
# Weergeef hun functie, en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden.

sales_staff = pd.read_sql_query(select_all_sales_staff, conn)
order_header = pd.read_sql_query(select_all_order_header, conn)

verkoper_is_manager = sales_staff[sales_staff['POSITION_EN'].str.contains('Manager')]

overzicht = verkoper_is_manager.merge(order_header, left_on = 'SALES_STAFF_CODE', how = 'left', right_on = 'SALES_STAFF_CODE')

resultaat = overzicht[['POSITION_EN', 'ORDER_DATE']].drop_duplicates()

resultaat





POSITION_EN  ORDER_DATE
0     Branch Manager         NaN
1   Regional Manager         NaN
4   District Manager         NaN
21    Branch Manager  2024-09-09
22    Branch Manager  2024-03-13
23    Branch Manager  2024-01-09
25   General Manager         NaN

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [90]:
# Welke tabellen? PRODUCT, PRODUCT_TYPE, ORDER_DETAILS
# Welke kolommen? 
# PRODUCT: PK: PRODUCT_NUMBER, PRODUCT_NAME FK: PRODUCT_TYPE_CODE
# PRODUCT_TYPE: PK: PRODUCT_TYPE_CODE, PRODUCT_TYPE_EN
# ORDER_DETAILS: PK: ORDER_DETAIL_CODE, QUANTITY
# Hoe joinen we deze?
# PRODUCT PRODUCT_TYPE_CODE, PRODUCT_TYPE_EN PRODUCT_TYPE_CODE, ORDER_DETAILS PRODUCT_NUMBER
# Lees de query's in
# Maak boolean, waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn.
# 2 Kolommen: naam van product, product_type.

product = pd.read_sql_query(select_all_product, conn)
product_type = pd.read_sql_query(select_all_product_type, conn)
order_details = pd.read_sql_query(select_all_order_details, conn)

overzicht = pd.merge(product, product_type, left_on = 'PRODUCT_TYPE_CODE', how = 'inner', right_on = 'PRODUCT_TYPE_CODE')

overzicht2 = pd.merge(overzicht, order_details, on = 'PRODUCT_NUMBER', how = 'inner')

product_750_tegelijk_verkocht = overzicht2['QUANTITY'] > 750

resultaat = overzicht2.loc[(product_750_tegelijk_verkocht, ['PRODUCT_NAME', 'PRODUCT_TYPE_EN'])].drop_duplicates()

resultaat




PRODUCT_NAME    PRODUCT_TYPE_EN
5427                Star Peg              Tents
11452              Firefly 2           Lanterns
28608        BugShield Spray  Insect Repellents
28892  BugShield Lotion Lite  Insect Repellents
29430       BugShield Lotion  Insect Repellents
29442      BugShield Extreme  Insect Repellents
30741         Sun Shelter 15          Sunscreen
31164         Sun Shelter 30          Sunscreen
31775             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [97]:
# Welke tabellen? PRODUCT, ORDER_DETAILS
# Welke kolommen?
# PRODUCT = PRODUCT_NAME
# ORDER_DETAILS = UNIT_PRICE, UNIT_SALE_PRICE
# Hoe joinen we deze?
# PRODUCT_NUMBER
# Lees de query's in
# Maak een boolean waarvan ooit meer dan 40% korting is verleend; dit doen we zo: '(unit_price - unit_sale_price / unit price) > 40
# Alleen de productnamen weergeven.

product = pd.read_sql_query(select_all_product, conn)
order_details = pd.read_sql_query(select_all_order_details, conn)

overzicht = pd.merge(product, order_details, on = 'PRODUCT_NUMBER', how = 'inner')

product_meer_dan_40_korting = ( (overzicht['UNIT_PRICE'] - overzicht['UNIT_SALE_PRICE']) / overzicht['UNIT_PRICE'] ) > 0.40

resultaat = overzicht.loc[(product_meer_dan_40_korting, ['PRODUCT_NAME'])].drop_duplicates()

resultaat 




PRODUCT_NAME
27830             BugShield Natural
28283               BugShield Spray
28694         BugShield Lotion Lite
29127              BugShield Lotion
29464             BugShield Extreme
30562                Sun Shelter 15
30929                Sun Shelter 30
34441  Hailstorm Titanium Woods Set

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [100]:
# Welke tabellen? RETURNED_ITEM, ORDER_DETAILS, RETURN_REASON
# Welke kolommen?
# RETURNED_ITEM : FK = ORDER_DETAIL_CODE, RETURN_QUANTITY
# ORDER_DETAILS : PK = ORDER_DETAIL_CODE, QUANTITY
# RETURN_REASON : PK : RETURN_REASON_CODE, RETURN_DESCRIPTION_EN
# Hoe joinen we deze?
# on = ORDER_DETAIL_CODE -> RETURN_REASON_CODE
# Lees de query's in
# Maak een boolean waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity)

returned_item = pd.read_sql_query(select_all_returned_item, conn)
order_details = pd.read_sql_query(select_all_order_details, conn)
return_reason = pd.read_sql_query(select_all_return_reason, conn)

overzicht = pd.merge(returned_item, order_details, on = 'ORDER_DETAIL_CODE', how = 'inner')

overzicht2 = pd.merge(overzicht, return_reason, on ='RETURN_REASON_CODE', how = 'inner')

retourde_items_meer_dan_90 = ( (overzicht2['RETURN_QUANTITY'] / overzicht2['QUANTITY']) ) > 0.90

resultaat = overzicht2.loc[retourde_items_meer_dan_90, ['RETURN_DESCRIPTION_EN']].drop_duplicates()

resultaat


RETURN_DESCRIPTION_EN
0  Unsatisfactory product
2   Wrong product shipped
3   Wrong product ordered

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (6 kolommen, 10 rijen)

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [2]:
from datetime import date
date.today().year

2026

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [3]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 30 jaar in dienst) of een ‘lang in dienst’ (groter dan gelijk aan 30 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 